keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1319 @@
+
+When the leader lease is marked as end of term or available, the coordinated leader election controller will
+add an annotation to all component leases `coordination.k8s.io/pending-ack` and wait up to 5 seconds. During that time, components must update their component lease to remove the annotation. The leader election controller will then pick the leader based on its criteria from the set of component leases that have ack'd the request.
+
+### Enabling on a component
+
+Components with a `--leader-elect-resource-lock` flag (kube-controller-manager,
+ kube-scheduler) will accept `coordinatedleases` as a resource lock type.
+
+### Migrations
+
+So long as the API server is running a coordinated election controller, it is
+safe to directly migrate a component from Lease Based Leader Election to
+Coordinated Leader Election (or vis-versa).
+
+During the upgrade, a mix of components will be running both election
+approaches. When the leader lease expires, there are a couple possibilities:
+
+- A controller instance using Lease Based Leader Election claims the leader
+  lease
+- The coordinated election controller picks a leader, from the components that
+  have written identity leases, and claims the lease on the leader's behalf
+
+Both possibilities have acceptable outcomes during the migration-- a component
+is elected leader, and once elected, remains leader so long as it keeps the
+lease renewed. The elected leader might not be the leader that Coordinated
+Leader Election would pick, but this is no worse than how leader election works
+before the upgrade, and once the upgrade is complete, Coordinated Leader
+Election works as intended.
+
+There is one thing that could make migrations slightly cleaner: If Coordinated
+Leader Election adds a `coordination.k8s.io/elected-by:
+leader-election-controller` annotation to any leases that it claims. It can also
+check for this annotation and only mark leases as "end-of-term" if that
+annotation is present. Lease Based Leader Election would ignore "end-of-term"
+annotations anyway, so this isn't strictly needed, but it would reduce writes
+from the coordinated election controller to leases that were claimed by
+component instances not using Coordinated Leader Election
+
+### API
+
+The lease lock API will be extended with a new field for election preference, denoted as an enum for strategies for Coordinated Leader Election.
+
+```go
+
+type CoordinatedLeaseStrategy string
+
+// CoordinatedLeaseStrategy defines the strategy for picking the leader for coordinated leader election.
+const (
+  OldestCompatibilityVersionStrategy CoordinatedStrategy = "OldestCompatibilityVersionStrategy"
+  NewestCompatibilityVersionStrategy CoordinatedStrategy = "NewestCompatibilityVersionStrategy"
+)
+
+type LeaseSpec struct {
+	// Strategy indicates the strategy for picking the leader for coordinated leader election
+	Strategy CoordinatedLeaseStrategy `json:"strategy,omitempty" protobuf:"string,6,opt,name=strategy"`
```

Comment:
As this is in the existing `Lease` type, I would expect a defaulted value for the non-coordindated case, e.g. `NoCoordination`.

Who creates the initial value in the coordinated case? The election coordinator has to get it from somewhere. Is that a value for the `LeaseCandidacy`? But then what if those values don't agree for all candidates? Do we need an order?

Reply:
Added a default value. We will create an ordering of values if they do not agree for all candidates.

Reply:
> Who creates the initial value in the coordinated case?

I was wondering the same. What actors in the system do we expect to be picking the `CoordinatedLeaseStrategy`?  If the controller implementation picks, then having the controllers write it to `IdentityLeases` and then deal with the case where the values don't agree for the candidates does seem like a possible approach.  It's more complex than I'd like though.

If the actor is not the controllers, what else would it be? I can imagine a election coordinator that has it's own configuration that it uses to make decisions, but that also starts to be significantly more complicated.